### ISSUE TITLE
Destinations V2: Only use SAFE_CAST if we have to

### ISSUE THREAD

**Issue by evantahler** [airbytehq/airbyte#30418]
I have a suspicion that our usage of SAFE_CAST is the most expensive.  I'd rather start investigating ways to maybe try a T&D without it first, and if it fails, try again with SAFE_CAST.  that might save a bunch of compute

This is a story in 2 parts:
1. Analyze if SAFE_CAST is expensive - run a (big) sync with and without it and get some data (BQ is good at logging cost information)
2. If it is effective, change our T&D code to try first without SAFE_CAST, and then only if there is a failure, try a second time with SAFE_CAST

```java
doTandD() {
  try {
    TandDUnsafely();
  } catch (BQException) {
    log(BQException);
    TandDWithSafeCast(); // what we have now
  }
}
```

**Comment by evantahler** [comment]
Gromming:

Some SQL dialects have an internal try/catch.  If BQ, Snowflake, Redshift, Postgres, MySQL, and Clickhouse all have this, :thumbsup:

**Comment by evantahler** [comment]
This change for unsafe-casts should also remove the array-building part of errors

**Comment by evantahler** [comment]
What's remaining on this story?

**Comment by evantahler** [comment]
At the moment - Done for Snowflake, Blocked for BigQuery.

**Comment by jbfbell** [comment]
more clarity on why this is Blocked:

in BigQuery, we were able to use BigQuery's native exception handling and did not have to change any interfaces `updateTable` just works as normal, we're adding more to the query, generally we have a pattern of:
create sql statement string (like via `updateTable`)
execute it
We had [something similar](https://github.com/airbytehq/airbyte/pull/31001/files) ready to go for Snowflake, but it fails on some probably real test edge cases. 

The TL;dr is that occasionally we might see a column name with a `$$` which is the same as Snowflake's Scripting "special quoting". I have tried other ways to slice this and use other quoting mechanisms to preserve the updateTable interface but something is getting lost in the escaping from java/JDBC -> snowflake where it just will not respect escaping `$$` in column names.
we currently do not have a notion of a Java implemented try/catch for SQL exceptions across the interfaces for DV2, so we have to change the interfaces to implement the try catch which isn't a extremely difficult change to make, maybe a day or two, but has a broader impact than changing the implementation of the snowflake specific `updateTable`.

Because there is other work in this exact area, we are blocking this just to avoid thrashing on the same file
